assets/shiny-server.js

```diff
@@ -637,6 +333,311 @@
+      def.reject();
+    });
+
+    return def;
+  }
+  // @param time The last time a connection attempt was started
+  // @param count 0-indexed count of how many reconnect attempts have occured.
+  // @param expires the time when the session is scheduled to expire on 
+  // the server.
+  MultiplexClient.prototype.scheduleReconnect_p = function(time, count, expires) {
+    var def = $.Deferred();
+
+    if (Date.now() > expires){
+      // Shouldn't happen, but if the current time is after the known
+      // expiration for this session, give up.
+      debug('Overshot session expiration.');
+      return def.reject(new Error("Overshot session expiration"));
+    }
+
+    // Compute delay exponentially.
+    var interval;
+    var delay;
+    if (count < 0){
+      interval = 0;
+      delay = 0;
+    } else {
+      if (count < 10){
+        interval = 1000 * Math.pow(2, count);
+        interval = Math.min(15 * 1000, interval); // Max of 15s delay.
+      } else { 
+        // The interval may end up being configurable or changed, so we don't cut off
+        // exactly at 2^4, but don't bother computing really large powers.
+        interval = 15 * 1000;
+      }
+      delay = time - Date.now() + interval;
+    }
+
+    // If the next attempt would be after the session is due to expire, 
+    // schedule one last attempt to connect a couple seconds before the
+    // expiration.
+    if (Date.now() + delay > (expires - 2000)) {
+      delay = expires - Date.now() - 2000;
+    }
+    if (delay < 0){
+      // i.e. we're within 2 seconds of session expiration. Make
+      // one last connection attmpt but don't schedule any more.
+      debug('Final reconnection attempt');
+      return this.reconnect_p();
+    }
+
+    var self = this;
+    function doRecon(){
+      var startTime = Date.now();
+      self.reconnect_p()
+      .then(function(c){
+        // Able to reconnect.
+        def.resolve(c);
+      }, function(){
+        self.scheduleReconnect_p(startTime, count+1, expires)
+        .then(function(c){
+          def.resolve(c);
+        }, function(e){
+          def.reject(e);
+        });
+      });
+    }
+
+    var reconTimeout = null;
+    this.setReconnectDialog('<button id="ss-reconnect-btn" type="button" class="ss-dialog-button">Reconnect Now</button>Trouble connecting to server.');
+    $('#ss-reconnect-btn').click(function(){
+      debug('Trying to reconnect now.');
+      // Short-circuit the wait.
+      clearTimeout(reconTimeout);
+      doRecon();
+    });
+    debug('Setting countdown timer');
+    debug('Scheduling reconnect attempt for ' + delay + 'ms');
+    // Schedule the reconnect for some time in the future.
+    reconTimeout = setTimeout(doRecon, delay);
+
+    return def;
+  }
+  MultiplexClient.prototype.startReconnect = function(){
+    var self = this;
+
+    // Schedule the display of the disconnect window
+    self._disconnectTimer = setTimeout(function(){
```

Comment:
Why are we starting the disconnect timer when we may have seen a clean disconnect? Should this be within the `self._disconnected` if statement?